source/pervasive/std_specs/num.rs

```diff
@@ -0,0 +1,281 @@
+    ensures 
+        rhs == 0 ==> result.is_None(),
+        rhs != 0 ==> {
+            match result {
+                Some(result) => result == lhs % rhs,
+                None => false 
+            }
+        }
+{
+    lhs.checked_rem_euclid(rhs)
+}
+
+// == i32 methods ==
+
+#[verifier::external_fn_specification]
+pub fn ex_i32_checked_add(lhs: i32, rhs: i32) -> (result: Option<i32>)
+    ensures 
+        lhs + rhs > i32::MAX || lhs + rhs < i32::MIN ==> result.is_None(),
+        i32::MIN <= lhs + rhs <= i32::MAX ==> {
+            match result {
+                Some(result) => result == lhs + rhs,
+                None => false
+            }
+        }
+{
+    lhs.checked_add(rhs)
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_i32_checked_add_unsigned(lhs: i32, rhs: u32) -> (result: Option<i32>)
+    ensures 
+        lhs + rhs > i32::MAX ==> result.is_None(),
+        lhs + rhs <= i32::MAX ==> {
+            match result {
+                Some(result) => result == lhs + rhs,
+                None => false
+            }
+        }
+{
+    lhs.checked_add_unsigned(rhs)
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_i32_checked_sub(lhs: i32, rhs: i32) -> (result: Option<i32>)
+    ensures 
+        lhs - rhs > i32::MAX || lhs - rhs < i32::MIN ==> result.is_None(),
+        i32::MIN <= lhs - rhs <= i32::MAX ==>
+            match result {
+                Some(result) => result == lhs - rhs,
+                None => false
+            }
+{
+    lhs.checked_sub(rhs)
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_i32_checked_sub_unsigned(lhs: i32, rhs: u32) -> (result: Option<i32>)
+    ensures
+        lhs - rhs < i32::MIN ==> result.is_None(),
+        i32::MIN <= lhs - rhs ==> 
+            match result {
+                Some(result) => result == lhs - rhs,
+                None => false
+            }
+{
+    lhs.checked_sub_unsigned(rhs)
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_i32_checked_mul(lhs: i32, rhs: i32) -> (result: Option<i32>)
+    ensures 
+        lhs * rhs < i32::MIN || lhs * rhs > i32::MAX ==> result.is_None(),
+        i32::MIN <= lhs * rhs <= i32::MAX ==>
+            match result {
+                Some(result) => result == lhs * rhs,
+                None => false 
+            }
+{
+    lhs.checked_mul(rhs)
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_i32_checked_div(lhs: i32, rhs: i32) -> (result: Option<i32>)
+    ensures 
+        rhs == 0 ==> result.is_None(),
+        ({
+            let x = lhs as int;
+            let d = rhs as int;
+            let output = if x == 0 {
+                0
+            } else if x > 0 && d > 0 {
+                x / d
+            } else if x < 0 && d < 0 {
+                ((x * -1) / (d * -1))
+            } else if x < 0 {
+                ((x * -1) / d) * -1
+            } else { // d < 0
+                (x / (d * -1)) * -1
+            };

```

Comment:
probably a good idea to factor the (lhs, rhs) -> output computation as a spec function on ints.